Support piped wikilinks

thomaskoppelaar opened the pull request:
Hello,

I've been tinkering around with this extension, and I think I've found a way to support piped wikilinks correctly:

- I've updated the regex for a Wikilink in src/NoteWorkspace.ts to allow for piped wikilinks
- And within src/Ref.ts, I've added a simple replace to remove wikilinks from the `word` attribute.

Please do check it out, I hope you find this useful!

thomaskoppelaar commented:
Edit: now that I think of it, I haven't taken a look at backlinks yet to see if they still get found properly - I'll mark this PR as a draft and check.

thomaskoppelaar commented:
This PR should now be good to go.

kortina commented on src/NoteWorkspace.ts:
```diff
@@ -171,6 +171,8 @@ export class NoteWorkspace {
     // remove the brackets:
     let n = noteName.replace(/[\[\]]/g, '');
+    // remove the potential wikilink:
+    n = n.replace(/\|[^\\\[]+$/, '');
```
OK, so am I correct in understanding the way this works is that the pipe and succeeding characters are stripped, such that:

```[[test-file.md|Some Test File]]``

Will resolve to the same thing as:

```[[test-file.md]]``

I can see this change affecting the following behaviors:

- `editor.action.revealDefinition` - *Go To Definition* on `[[test-file.md|Some Test]]` should open `test-file.md` (if `test-file.md` exists on file system)
- `createNoteOnGoToDefinitionWhenMissing` - `editor.action.revealDefinition`(missing *Go To Definition*) on `[[test-file.md|Some Test]]` should CREATE and open `test-file.md` (if `test-file.md` does NOT exist on file system)
- `editor.action.peekDefinition` - should preview `test-file.md` (if `test-file.md` does exists on file system)
- references / backlinks `[[test-file.md|Some Test]]` should register as a reference / backlink to `test-file.md`

### questions

- Does your PR address all of these behaviors, or just some of them?
- Can you please add some test cases involving piped links in https://github.com/kortina/vscode-markdown-notes/blob/master/src/test/jest/extension.test.ts
- Can you feature flag this with a new config setting, `vscodeMarkdownNotes.allowPipedWikiLinks`, where if this is true we strip the pipe and succeeding characters, otherwise treat the pipe as any other special character that gets replaced with the `slugifyCharacter`
- Perhaps we should make the pipe character itself configurable, defaulting it to "|" ? @lukesmurray suggested making this configurable elsewhere.
- Do Obsidian, Roam Research, Foam, and other projects all treat the characters before the pipe as the file and the characters after as the link text (or does anyone do the inverse)? Seems like your PR follows the [mediawiki style](https://www.mediawiki.org/wiki/Help:Links/sco) and I just want to make sure this how it works for other projects.

Thank you for the PR! I am excited to add this functionality, as several other users of the extension have asked for it.

Lmk if you have any questions about how to proceed with the PR by addressing the feedback.

---

This PR will address #56 and #66

thomaskoppelaar commented:
I've quickly tested all of the behavior you mentioned - There shouldn't be any problem occurring with the way I've implemented this.

I've also added two new settings: `allowPipedWikiLinks`, and `pipedWikiLinksSyntax`.

One question: Should the Ref object contain the description, in case it exists? I don't see any particular use for this, but I guess that it could be done in order to give the object a little more data.

thomaskoppelaar commented:
@kortina Requesting a review! LMK if you find anything out of the ordinary or broken, or if you need me to test something more thoroughly.

kortina commented:
Re: "Should the Ref object contain the description, in case it exists? I don't see any particular use for this, but I guess that it could be done in order to give the object a little more data." - I don't nec think so, because you'll be able to see the entire line in the preview of the file, so doesn't seem terribly additive.

Open to other perspectives / and a good question tho. I am not opposed to it and could see it maybe being useful as part of the API / for other people working on the extension...

kortina commented on package.json:
```diff
@@ -94,6 +94,27 @@
+          "type": "string",
+          "enum": [
+            "desc|file",
+            "file|desc"
```
Hey, I am guessing you added this setting in response to the question:

> Do Obsidian, Roam Research, Foam, and other projects all treat the characters before the pipe as the file and the characters after as the link text (or does anyone do the inverse)? Seems like your PR follows the mediawiki style and I just want to make sure this how it works for other projects.

I am curious what you found here -- were there some programs that inverted the order here? Did you find any consensus on this?

If there are different ways it works out there, I think the setting is the way to handle.

If all the related programs handle consistently, maybe the setting is just confusing.

Any thoughts on this before we merge?

thomaskoppelaar replied:
> I am curious what you found here -- were there some programs that inverted the order here? Did you find any consensus on this?

Two things:  [MediaWiki's meta page](https://meta.wikimedia.org/wiki/Help:Piped_link) and [What Github Wiki uses](https://github.com/gollum/gollum/wiki#link-tag)

> If there are different ways it works out there, I think the setting is the way to handle.

Yes, agreed.

kortina commented on src/NoteWorkspace.ts:
```diff
@@ -41,7 +49,7 @@ export class NoteWorkspace {
   static _rxTagNoAnchors = '\\#[\\w\\-\\_]+'; // used to match tags that appear within lines
   static _rxTagWithAnchors = '^\\#[\\w\\-\\_]+$'; // used to match entire words
-  static _rxWikiLink = '\\[\\[[^\\]]+\\]\\]'; // [[wiki-link-regex]]
+  static _rxWikiLink = "\\[\\[[^sep\\]]+(sep[^sep\\]]+)?\\]\\]"; // [[wiki-link-regex(|with potential pipe)?]]
```
Can you add a line comment that notes that `sep` will be replaced with `pipedWikiLinksSeparator` in this regex on compile?

thomaskoppelaar replied:
Done!

kortina commented on src/test/jest/extension.test.ts:
```diff
@@ -124,6 +128,28 @@ test('noteNamesFuzzyMatchSlashes', () => {
+    "file"
+  );
+
+  // Odd case, but I suppose it should be treated
```
If nothing else, it serves as clarification of behavior.

kortina commented on src/test/jest/extension.test.ts:
```diff
@@ -8,7 +8,11 @@ jest.mock('../../NoteWorkspace');
 beforeEach(() => {
   NoteWorkspace.cfg = () => {
-    return NoteWorkspace.DEFAULT_CONFIG;
+    let config =  NoteWorkspace.DEFAULT_CONFIG;
```
If we are keeping the `pipedWikiLinksSyntax` can you add some test cases for the non-default config?

Also may be worth having a test for use of something else as `pipedWikiLinksSeparator` as well as an invalid `pipedWikiLinksSeparator`

thomaskoppelaar replied:
> If we are keeping the pipedWikiLinksSyntax can you add some test cases for the non-default config?
Sure - I sadly don't have time to do this tonight, I'll get to it tomorrow.

> Also may be worth having a test for use of something else as `pipedWikiLinksSeparator`
Will do!

> as well as an invalid `pipedWikiLinksSeparator`

This is something we can prevent from happening by using a good pattern within `package.json`:

```json
{
    "pattern": "[^\\[\\]]+",
    "patternErrorMessage": "Warning: This separator will unfortunately not work - It would break the wikilink syntax."
}
```
This blocks a user from ever setting "[" or "]" as a separator.
